typing/ctype.ml

```diff
@@ -263,6 +263,36 @@ let newconstr path tyl = newty (Tconstr (path, tyl, ref Mnil))
 
 let none = newty (Ttuple [])                (* Clearly ill-formed type *)
 
+(**** information for [Typecore.unify_pat_*] ****)
+
+module Pattern_env : sig
+  type env_ref
+  type t = private
+    { renv : env_ref;
```

Comment:
May I ask why your hiding the env behind an abstract type?
This is really nitpicking on my part, but it's adding an indirection when we could just mark the field as mutable.

Reply:
I think that making the field mutable would not quite work with the current API, because `set_equations_scope : int -> t -> t` would then copy the mutable field instead of sharing the mutable reference (which I think is the desired behavior here). It is rather tricky what when we want to share or copy, and I would guess that an `env ref` field is necessary. I think that hiding `env ref` under and abstraction barrier is not strictly necessary, but I am okay with it.

Reply:
The uses of the reference turned out to be indeed unnecessary and we removed them. Could you read the updated code again?